src/Drips.sol

```diff
@@ -377,113 +403,167 @@ library Drips {
-                // all matching pairs of drips receiver configurations will be found.
-                uint256 currReceiver = currReceivers[currIdx].userId;
-                uint256 newReceiver = newReceivers[newIdx].userId;
-                pickCurr = currReceiver <= newReceiver;
-                pickNew = newReceiver <= currReceiver;
+            DripsReceiver memory newRecv;
+            if (pickNew) newRecv = newReceivers[newIdx];
+
+            // Limit picking both curr and new to situations when they differ only by time
+            if (
+                pickCurr &&
+                pickNew &&
+                (currRecv.userId != newRecv.userId || currRecv.amtPerSec != newRecv.amtPerSec)
+            ) {
+                pickCurr = _isOrdered(currRecv, newRecv);
+                pickNew = !pickCurr;
             }
-            // The drips update parameters
-            uint256 receiver;
-            int128 currAmtPerSec = 0;
-            int128 newAmtPerSec = 0;
-            if (pickCurr) {
-                receiver = currReceivers[currIdx].userId;
-                currAmtPerSec = int128(currReceivers[currIdx].amtPerSec);
-                // Clear the obsolete drips end
-                _setDelta(
-                    s.dripsStates[receiver][assetId].amtDeltas,
-                    cycleSecs,
-                    currEndTime,
-                    currAmtPerSec
+
+            if (pickCurr && pickNew) {
+                // Shift the existing drip to fulfil the new configuration
+                mapping(uint64 => AmtDelta) storage deltas = s
+                .dripsStates[currRecv.userId][assetId].amtDeltas;
+                (uint64 currStart, uint64 currEnd) = _dripsRange(
+                    currRecv,
+                    lastUpdate,
+                    currDefaultEnd
                 );
-                currIdx++;
-            }
-            if (pickNew) {
-                receiver = newReceivers[newIdx].userId;
-                newAmtPerSec = int128(newReceivers[newIdx].amtPerSec);
-                // Apply the new drips end
-                _setDelta(
-                    s.dripsStates[receiver][assetId].amtDeltas,
-                    cycleSecs,
-                    newEndTime,
-                    -newAmtPerSec
+                (uint64 newStart, uint64 newEnd) = _dripsRange(
+                    newRecv,
+                    _currTimestamp(),
+                    newDefaultEnd
                 );
-                newIdx++;
-            }
-            // The receiver may have never been used
-            if (!pickCurr) {
-                DripsState storage dripsState = s.dripsStates[receiver][assetId];
-                // The receiver has never been used, initialize it
+                _clearDeltaRange(deltas, cycleSecs, currStart, newStart, currRecv.amtPerSec);
+                _setDeltaRange(deltas, cycleSecs, currEnd, newEnd, currRecv.amtPerSec);
```

Comment:
I think we should make the `_setDeltaRange` here twice. Using the `_clearDeltaRange` here is confusing.

Reply:
But then we'd need to switch `currStart` and `newStart`, IMO that'd be weird and more difficult to visualize. If you have a strong opinion on that, I'm ok with changing that.

Reply:
It depends on which layer you want to understand it. 
For me it caused confusion because i didn't understand why both start times are passed. 

Maybe at this level it would be more clear with something like this?

```solidity
// move start and end only if it has changed
moveDelta(currStart, currEnd, newStart, newEnd, currRecv.amtPerSec);
```
If it doesn't produce any additional gas.

Reply:
Ok, that's good. I'm quite certain that it'll be inlined. Added.